llvm/lib/Target/NVPTX/SYCL/GlobalOffset.cpp

```diff
@@ -0,0 +1,368 @@
+//===--------- GlobalOffset.cpp - Global Offset Support for CUDA --------- ===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This pass operates on SYCL kernels being compiled to CUDA. It looks for uses
+// of the `llvm.nvvm.implicit.offset` intrinsic and replaces it with a offset
+// parameter which will be threaded through from the kernel entry point.
+//
+//===----------------------------------------------------------------------===//
+
+#include "GlobalOffset.h"
+
+#include "../MCTargetDesc/NVPTXBaseInfo.h"
+#include "llvm/ADT/SmallSet.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Intrinsics.h"
+#include "llvm/IR/IntrinsicsNVPTX.h"
+#include "llvm/IR/PassManager.h"
+#include "llvm/Target/TargetIntrinsicInfo.h"
+#include "llvm/Transforms/Utils/Cloning.h"
+
+using namespace llvm;
+
+#define DEBUG_TYPE "globaloffset"
+
+namespace llvm {
+void initializeGlobalOffsetPass(PassRegistry &);
+} // end namespace llvm
+
+namespace {
+
+class GlobalOffset : public ModulePass {
+public:
+  static char ID;
+  GlobalOffset() : ModulePass(ID) {}
+
+  bool runOnModule(Module &M) override {
+    if (skipModule(M))
+      return false;
+
+    llvm::Function *ImplicitOffsetIntrinsic =
+        M.getFunction(Intrinsic::getName(Intrinsic::nvvm_implicit_offset));
+    KernelImplicitArgumentType =
+        ArrayType::get(Type::getInt32Ty(M.getContext()), 3);
+    ImplicitOffsetPtrType = Type::getInt32Ty(M.getContext())->getPointerTo();
+    assert(
+        (!ImplicitOffsetIntrinsic ||
+         ImplicitOffsetIntrinsic->getReturnType() == ImplicitOffsetPtrType) &&
+        "Intrinsic::nvvm_implicit_offset does not return the expected "
+        "type");
+
+    if (!ImplicitOffsetIntrinsic || ImplicitOffsetIntrinsic->use_empty()) {
+      return false;
+    }
+
+    // Find all entry points.
+    EntryPointMetadata = getEntryPointMetadata(M);
+
+    // Add implicit parameters to all direct and indirect users of the offset
+    this->addImplicitParameterToCallers(M, ImplicitOffsetIntrinsic, nullptr);
+
+    // Assert that all uses of `ImplicitOffsetIntrinsic` are removed and delete
+    // it.
+    assert(ImplicitOffsetIntrinsic->use_empty() &&
+           "Not all uses of intrinsic removed");
+    ImplicitOffsetIntrinsic->eraseFromParent();
+
+    return true;
+  }
+
+  void processKernelEntryPoint(Module &M, Function *Func) {
+    assert(EntryPointMetadata.count(Func) != 0 &&
+           "Function must be an entry point");
+
+    LLVMContext &Ctx = M.getContext();
+    MDNode *FuncMetadata = EntryPointMetadata[Func];
+
+    bool AlreadyProcessed = this->ProcessedFunctions.count(Func) == 1;
```

Comment:
```suggestion
    bool AlreadyProcessed = ProcessedFunctions.count(Func) == 1;
```
Just out of curiosity: Is accessing members though `this->` really necessary in this pass?

Reply:
I've added a TODO comment, but I wonder how to make sure that the PI CUDA backend would be able to easily find the new kernels if the naming changes.

Reply:
I think it's old habits, but you're absolutely right that they're all unnecessary so I've removed them.